design/restore-with-EnableAPIGroupVersions-feature.md

```diff
@@ -0,0 +1,211 @@
+  
+  Once the version to restore is chosen, the file path to the backed up resource in the tarball will be modified such that it points to the resources' chosen API group version. If no version is found in common between the source and target clusters, the chosen version will default to the source cluster's preferred version (the version being restored currently without the changes proposal here). Restore will be allowed to continue as before.
+
+## Detailed Design
+
+There are six objectives to achieve the above stated goals:
+
+1. Determine if the APIGroupVersionsFeatureFlag feature flag is enabled and Backup Objects use Status.FormatVersion 1.1.0.
+1. List the backed up API group versions.
+1. List the API group versions supported by the target cluster.
+1. Get the user-defined version priorities.
+1. Use a priority system to determine which version to restore. The source preferred version will be the default if the priorities fail.
+1. Modify the paths to the backup files in the tarball in the resource restore process.
+
+### Objective 1: Determine if the APIGroupVersionsFeatureFlag feature flag is enabled and Backup Objects use Status.FormatVersion 1.1.0
+
+For restore to be able to choose from multiple supported backed up versions, the feature flag must have been enabled during the restore processes. Backup objects must also have [Status.FormatVersion == "1.1.0"](https://github.com/vmware-tanzu/velero/blob/a1e182e723a8c5f6d4175d8db2361233a94d2502/pkg/backup/backup.go#L58).
+
+The reason for checking for the feature flag during restore is to ensure the user would like to restore a version that might not be the source cluster preferred version.
+
+The reason for checking `Status.FormatVersion` is to ensure the changes made by this proposed design is backward compatible. Only with Velero version 1.4 and forward was Format Version 1.1.0 used to structure the backup directories. Format Version 1.1.0 is required for the restore process proposed in this design doc to work. Before v1.4, the backed up files were in a directory structure that will not be recognized by the proposed code changes. In this case, restore should not attempt to restore from multiple versions as they will not exist.
+
+Checking if the feature flag is enabled during restore is straightforward using `features.IsEnabled(velerov1api.APIGroupVersionsFeatureFlag)`.
+
+Checking the `Status.FormatVersion` can be done indirectly by checking the directory structure and version directory names. For backups generated using Format Version 1.1.0, there will always be a version directory with `-preferredversion` in the directory name. Only one resource will need to be checked to determine if the feature flag was enabled during backup.
```

Comment:
Can this paragraph be deleted, since on L57 you're looking directly at the backup object's `Status.FormatVersion` field? Or is this a double check to be sure?

Reply:
To clarify - [the Backup object is included in a restore Request](https://github.com/vmware-tanzu/velero/blob/main/pkg/restore/restore.go#L81).

It is [also on the restoreContext](https://github.com/vmware-tanzu/velero/blob/main/pkg/restore/restore.go#L229).

So we have direct access to the Velero `Backup` CR that's being restored, not just the tarball.

Reply:
Oh, I think I misread something you wrote earlier. I do see `status.formatVersion` inside `velero-backup.json`. Would this be where I grabbed the format version from?

And if I can do that, I definitely won't need to be looking through the directories to look for `-preferredversion` in the names.

Reply:
I think in this case, it would end up being from the restoreContext.

Something like `ctx.backup.Status.FormatVersion`.

You shouldn't have to parse the `velero-backup.json` yourself, as that's done for you. The JSON file will end up in the `ctx.backup` variable as a full Go struct.

Reply:
Nice! Thanks so much @nrb. I wrote my last question after your clarification. I'll fix the paragraphs to use the backup info in the `restoreContext`. That's much better.

Reply:
The `Backup` comes from [this location](https://github.com/vmware-tanzu/velero/blob/6808acd92e30848056a21faf373af03ddb8a3b71/pkg/controller/restore_controller.go#L403) up in the restore controller, as an FYI. :)

So it's not directly from the JSON file, but rather the in-cluster representation. The JSON file is used when Velero syncs into brand new clusters, though.